src/mlpack/methods/lsh_model/lshmodel.hpp

```diff
@@ -0,0 +1,373 @@
+    * are specified, we will not used the trained but the provided parameters.
+    * If these are left to default (0), the estimated parameters will be used.
+    *
+    * @param numProjIn The number of projections per table.
+    * @param numTablesIn The number of projection tables.
+    * @param hashWidthIn The first level hash width.
+    * @param secondHashSize The second level hash width.
+    * @param bucketSize The second level bucket size.
+    */
+   LSHSearch<SortPolicy> LSHObject(
+       const size_t numProjIn = 0,
+       const size_t numTablesIn = 0,
+       const double hashWidthIn = 0.0,
+       const size_t secondHashSize = 99901,
+       const size_t bucketSize = 500);
+
+   //! Return the number of projections calculated.
+   size_t NumProj(void) const { return numProj; };
+
+   //! Return the number of tables calculated.
+   size_t NumTables(void) const { return numTables; };
+
+   //! Return the calculated hash width.
+   double HashWidth(void) const { return hashWidth; };
+
+   //! Return the calculated number of probes.
+   double NumProbes(void) const { return numProbes; };
+
+   //! Return the reference set.
+   const arma::mat ReferenceSet(void) const {return *referenceSet; };
+
+   //! Serialize the LSHModel object.
+   template<typename Archive>
+   void Serialize(Archive& ar);
+
+ private:
+
+  /**
+   * Returns the score of a perturbation vector generated by perturbation set A.
+   * The score of a pertubation set (vector) is the sum of scores of the
+   * participating actions.
+   * @param A perturbation set to compute the score of.
+   * @param scores vector containing score of each perturbation.
+  */
+  double PerturbationScore(const std::vector<bool>& A,
+                                  const arma::vec& scores) const;
+  /**
+   * Inline function used by GetAdditionalProbingBins. The vector shift operation
+   * replaces the largest element of a vector A with (largest element) + 1.
+   * Returns true if resulting vector is valid, otherwise false.
+   * @param A perturbation set to shift.
+  */
+  bool PerturbationShift(std::vector<bool>& A) const;
+
+  /**
+   * Inline function used by GetAdditionalProbingBins. The vector expansion
+   * operation adds the element [1 + (largest_element)] to a vector A, where
+   * largest_element is the largest element of A. Returns true if resulting vector
+   * is valid, otherwise false.
+   * @param A perturbation set to expand.
+  */
+  bool PerturbationExpand(std::vector<bool>& A) const;
+
+  /**
+   * Return true if perturbation set A is valid. A perturbation set is invalid if
+   * it contains two (or more) actions for the same dimension or dimensions that
+   * are larger than the queryCode's dimensions.
+   * @param A perturbation set to validate.
+   * @param numProj The number of projections for the sequence under validation.
+  */
+  bool PerturbationValid(const std::vector<bool>& A, size_t numProj) const;
```

Comment:
I believe these Perturbation* functions are exactly the same as in `LSHSearch`, so maybe we should mark those methods static and use them here, or create some auxiliary class for multi-probe LSH searches and use that.